streams/src/main/java/org/apache/kafka/streams/state/internals/LogicalSegmentIterator.java

```diff
@@ -100,34 +134,49 @@ private boolean maybeFillIterator() {
                 } else {
-                    // this segment contains records with the specified key and time range
-                    final List<RocksDBVersionedStoreSegmentValueFormatter.SegmentValue.SegmentSearchResult> searchResults =
-                            RocksDBVersionedStoreSegmentValueFormatter.deserialize(rawSegmentValue).findAll(fromTime, toTime);
-                    for (final RocksDBVersionedStoreSegmentValueFormatter.SegmentValue.SegmentSearchResult searchResult : searchResults) {
-                        queryResults.add(new VersionedRecord<>(searchResult.value(), searchResult.validFrom(), searchResult.validTo()));
-                    }
+                    this.currentDeserializedSegmentValue = RocksDBVersionedStoreSegmentValueFormatter.deserialize(rawSegmentValue);
+                    this.minTimestamp = RocksDBVersionedStoreSegmentValueFormatter.getMinTimestamp(rawSegmentValue);
+                    this.nextTimestamp = RocksDBVersionedStoreSegmentValueFormatter.getNextTimestamp(rawSegmentValue);
                 }
-            }
-            if (!queryResults.isEmpty()) {
-                break;
+                return true;
             }
         }
-        if (!queryResults.isEmpty()) {
-            // since data is stored in descending order in the segments, create the list in reverse order, if the order is Ascending.
-            this.iterator = order.equals(ResultOrder.ASCENDING) ? queryResults.listIterator(queryResults.size()) : queryResults.listIterator();
-            return true;
-        }
         // if all segments have been processed, release the snapshot
         releaseSnapshot();
         return false;
     }
 
+    private Object getNextRecord() {
+        VersionedRecord nextRecord = null;
+        if (currentRawSegmentValue != null) { // this is the latestValueStore
+            final long recordTimestamp = RocksDBVersionedStore.LatestValueFormatter.getTimestamp(currentRawSegmentValue);
+            if (recordTimestamp <= toTime) {
+                final byte[] value = RocksDBVersionedStore.LatestValueFormatter.getValue(currentRawSegmentValue);
+                // latest value satisfies timestamp bound
+                nextRecord = new VersionedRecord<>(value, recordTimestamp);
+            }
+        } else {
+            final RocksDBVersionedStoreSegmentValueFormatter.SegmentValue.SegmentSearchResult currentRecord =
+                    currentDeserializedSegmentValue.find(fromTime, toTime, order);
+            if (currentRecord != null) {
+                nextRecord = new VersionedRecord<>(currentRecord.value(), currentRecord.validFrom(), currentRecord.validTo());
+            }
+        }
+        // no relevant record can be found in the segment
+        if (currentRawSegmentValue != null || nextRecord == null || !canSegmentHaveMoreRelevantRecords(nextRecord.timestamp(), Long.parseLong(nextRecord.validTo().get().toString()))) {
```

Comment:
`Long.parseLong(nextRecord.validTo().get().toString())` -- why to we get valid-to as `String` and pipe though `parseLong`? Why can't we just pass `nextRecord.validTo().get()` into `canSegmentHaveMoreRelevantRecords(...)`?

Reply:
> `Long.parseLong(nextRecord.validTo().get().toString())` -- why to we get valid-to as `String` and pipe though `parseLong`? Why can't we just pass `nextRecord.validTo().get()` into `canSegmentHaveMoreRelevantRecords(...)`?

I had to do that. Fixing generic type, fixed all these issues now :)